router/data_store.go

```diff
@@ -39,196 +33,315 @@ type SyncHandler interface {
-	keys = newKeys
-	newKeys = make(map[string]uint64)
 
-watch:
-	if started != nil {
-		started <- nil
-		started = nil
+	if err != nil {
+		log.Printf("router: datastore error: %s, %s", id, err)
+		return
 	}
 
-	for {
-		stream := make(chan *etcd.Response)
-		watchDone := make(chan struct{})
-		var watchErr error
-		go func() {
-			_, watchErr = s.etcd.Watch(s.prefix, nextIndex, true, stream, s.stopSync)
-			close(watchDone)
-		}()
-		for res := range stream {
-			nextIndex = res.EtcdIndex + 1
-			id := path.Base(res.Node.Key)
-			var err error
-			if res.Action == "delete" {
-				delete(keys, id)
-				err = h.Remove(id)
-			} else {
-				keys[id] = res.Node.ModifiedIndex
-				route := &router.Route{}
-				if err = json.Unmarshal([]byte(res.Node.Value), route); err != nil {
-					goto fail
-				}
-				err = h.Set(route)
+	if err := h.Set(route); err != nil {
+		log.Printf("router: sync handler set error: %s, %s", id, err)
+	}
+}
+
+func (d *pgDataStore) startListener(idc chan<- string) error {
+	conn, err := d.pgx.Acquire()
+	if err != nil {
+		return err
+	}
+	if err = conn.Listen(d.tableName); err != nil {
+		d.pgx.Release(conn)
+		return err
+	}
+
+	go func() {
+		defer unlistenAndRelease(d.pgx, conn, d.tableName)
+		defer close(idc)
+
+		for {
+			select {
+			case <-d.donec:
+				return
+			default:
+			}
+			notification, err := conn.WaitForNotification(time.Second)
+			if err == pgx.ErrNotificationTimeout {
+				continue
 			}
-		fail:
 			if err != nil {
-				log.Printf("Error while processing update from etcd: %s, %#v", err, res.Node)
+				log.Printf("router: notifier error: %s", err)
+				d.StopSync()
+				return
 			}
+
+			idc <- notification.Payload
 		}
-		<-watchDone
-		select {
-		case <-s.stopSync:
-			return
-		default:
-		}
-		if e, ok := watchErr.(*etcd.EtcdError); ok && e.ErrorCode == 401 {
-			// event log has been pruned beyond our waitIndex, force fullSync
-			log.Printf("Got etcd error 401, doing full sync")
-			goto fullSync
-		}
-		log.Printf("Restarting etcd watch %s due to error: %s", s.prefix, watchErr)
-		time.Sleep(time.Second)
+	}()
+
+	return nil
+}
+
+const (
+	selectColumnsHTTP = "id, parent_ref, service, domain, sticky, tls_cert, tls_key, created_at, updated_at"
+	selectColumnsTCP  = "id, parent_ref, service, port, created_at, updated_at"
+)
+
+func (d *pgDataStore) columnNames() string {
+	switch d.routeType {
+	case "http":
+		return selectColumnsHTTP
+	case "tcp":
+		return selectColumnsTCP
+	default:
+		panic(fmt.Sprintf("unknown routeType: %q", d.routeType))
+	}
+}
+
+type scannable interface {
+	Scan(dest ...interface{}) (err error)
+}
+
+func (d *pgDataStore) scanRoute(route *router.Route, s scannable) error {
+	route.Type = d.routeType
+	switch d.tableName {
+	case "http_routes":
+		return s.Scan(
+			&route.ID,
+			&route.ParentRef,
+			&route.Service,
+			&route.Domain,
+			&route.Sticky,
+			&route.TLSCert,
+			&route.TLSKey,
+			&route.CreatedAt,
+			&route.UpdatedAt,
+		)
+	case "tcp_routes":
```

Comment:
yeah, good call, didn't even realize how much these were being repeated. I'm also changing the `routeType` values to constants (`http` and `tcp`).